pkg/operator/metrics/metrics.go

```diff
@@ -0,0 +1,161 @@
+// prometheus cardinality this set of outcomes should be kept small and only used for coarse and very high value
+// categories. Controllers must report an outcome but can report unspecified if this categorization is not needed.
+type ReconcileOutcome string
+
+const (
+	ReconcileOutcomeUnspecified        ReconcileOutcome = "unspecified"
+	ReconcileOutcomeNoOp               ReconcileOutcome = "no-op"
+	ReconcileOutcomeSkippedSync        ReconcileOutcome = "skipped-sync"
+	ReconcileOutcomeFullSync           ReconcileOutcome = "full-sync"
+	ReconcileOutcomeClusterSyncCreated ReconcileOutcome = "clustersync-created"
+)
+
+func init() {
+	metrics.Registry.MustRegister(metricHiveConfigConditions)
+	metrics.Registry.MustRegister(metricControllerReconcileTime)
+}
+
+// Add creates a new metrics Calculator and adds it to the Manager.
+func Add(mgr manager.Manager) error {
+	mc := &Calculator{
+		Client:   mgr.GetClient(),
+		Interval: 2 * time.Minute,
+	}
+	err := mgr.Add(mc)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Calculator runs in a goroutine and periodically calculates and publishes
+// Prometheus metrics which will be exposed at our /metrics endpoint. Note that this is not
+// a standard controller watching Kube resources, it runs periodically and then goes to sleep.
+//
+// This should be used for metrics which do not fit well into controller reconcile loops,
+// things that are calculated globally rather than metrics related to specific reconciliations.
+type Calculator struct {
+	Client client.Client
+
+	// Interval is the length of time we sleep between metrics calculations.
+	Interval time.Duration
+}
+
+// Start begins the metrics calculation loop.
+func (mc *Calculator) Start(ctx context.Context) error {
+	log.Info("started metrics calculator goroutine")
+
+	// Run forever, sleep at the end:
+	wait.UntilWithContext(ctx, func(ctx context.Context) {
+		mcLog := log.WithField("controller", "metrics")
+		recobsrv := NewReconcileObserver(ControllerName, mcLog)
+		defer recobsrv.ObserveControllerReconcileTime()
+
+		mcLog.Info("calculating metrics for all Hive")
+		// Load all ClusterDeployments so we can accumulate facts about them.
+		hiveConfig := &hivev1.HiveConfig{}
+
+		err := mc.Client.Get(context.TODO(), types.NamespacedName{Name: constants.HiveConfigName}, hiveConfig)
+		if err != nil {
+			mcLog.WithError(err).Error("error reading HiveConfig")
+			return
+		}
+
+		mc.setHiveConfigMetrics(hiveConfig)
+
+	}, mc.Interval)
+
+	return nil
+}
+
+func (mc *Calculator) setHiveConfigMetrics(config *hivev1.HiveConfig) {
+	for _, condition := range config.Status.Conditions {
+		if condition.Status == corev1.ConditionTrue {
+			metricHiveConfigConditions.WithLabelValues(string(condition.Type), condition.Reason).Set(1)
+		} else {
+			metricHiveConfigConditions.WithLabelValues(string(condition.Type), condition.Reason).Set(0)
+		}
+	}
+}
```

Comment:
So we don't ever expect a Condition to disappear now I guess, if it did we'd just keep reporting the previous value until the pod was restarted. This is probably no longer a concern though.

Reply:
If we want to have this metric displayed as a table then may be we should clear the history and only report latest status per condition WDYT?

Reply:
if the conditions are always present, be then unknown true or false, then it should effectively clear itself as it is now. If we had conditions literally going away then we'd have a problem, but it looks like it should be ok as is to me.

Reply:
(Clearing gets a little tricky as we'd need to zero out any possible combo of condition + reason to get rid of all the time series.)